### PR TITLE
Add Tag on task using Enter

### PR DESCRIPTION
**Pull request by panoramix360**
This PR adds the functionality to select Tags when the user presses Enter while searching for tags.

It will always pick the first tag that appears on the list, if there are no tags, nothing will be added.

The behavior is detailed on [the issue](https://github.com/dwyl/mvp/issues/308).

I tried to make it simple.

Please, let me know if you guys think it's needed any changes.

I didn't like the way I managed to empty the field after the user press Enter but I tried using LiveView <.input> with no success, if you guys have any other ideas let me know, I'd love to learn since I'm new to Phoenix/LiveView.

:D

### CODE REVIEW

**Comment by panoramix360** [comment]
hey @nelsonic, I notice that the Github Checks didn't passed, is there something on my end that I need to fix?

**Comment by LuchoTurtle** [comment]
Hey @panoramix360 .
Thanks for the PR! Run `mix format` to fix the `Build and Test` workflow. We have it setup so the code is correctly formatted before being pushed into the repo. 
Thanks!

**Comment by panoramix360** [comment]
thanks @LuchoTurtle!

Done!

**Comment by LuchoTurtle** [comment]
@nelsonic we're getting an error with the deploy review app script.

https://github.com/dwyl/mvp/actions/runs/5488807164/jobs/10002305167

```
Review App Script
Error: No access token available. Please login with 'flyctl auth login'
```

Have you stumbled upon this error before?

**Comment by LuchoTurtle** [comment]
We need to set `FLY_ACCESS_TOKEN` in order for this to work.

https://fly.io/docs/flyctl/integrating/

**Comment by panoramix360** [comment]
@LuchoTurtle, is this something that I should do on the PR?

File: lib/app_web/live/app_live.ex
Comment (LuchoTurtle) [inline lib/app_web/live/app_live.ex:106]: Can you place this function not in-between `handle_event` functions? So the file is correctly formatted, `mix format` is tossing this error:

```
clauses with the same name and arity (number of arguments) should be grouped together, "def handle_event/3" was previously defined (lib/app_web/live/app_live.ex:45)
```

This can easily be solved by just placing the private function alongside the others and let the `handle_event` functions be grouped together 😄

**Review by LuchoTurtle — CHANGES_REQUESTED** [review]

File: lib/app_web/live/app_live.ex
Comment (panoramix360) [reply lib/app_web/live/app_live.ex]: You are right, sorry for that! it's fixed!

**Comment by nelsonic** [comment]
@panoramix360 I think the CI doesn't run because you created a fork instead of a branch. 💭 
We need to fix that on our side as @LuchoTurtle says. 🔧 
You have an invite for `write` access to the repo for any future contributions. ✍️

**Comment by nelsonic** [comment]
I checked out @panoramix360 branch:
```sh
git clone git@github.com:panoramix360/mvp.git && cd mvp
git checkout 'feature/add-task-with-tag-when-enter'
cp ../.env .
source .env
mix deps.get
mix c
```

Got:

```sh
Finished in 1.2 seconds (1.1s async, 0.1s sync)
118 tests, 0 failures

Randomized with seed 640070
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/api/item.ex                               218       56        0
100.0% lib/api/tag.ex                                101       24        0
100.0% lib/api/timer.ex                              152       40        0
100.0% lib/app/color.ex                               90        1        0
100.0% lib/app/item.ex                               377       53        0
100.0% lib/app/item_tag.ex                            12        1        0
100.0% lib/app/tag.ex                                108       18        0
100.0% lib/app/timer.ex                              481       90        0
100.0% lib/app_web/controllers/auth_controller.       26        4        0
100.0% lib/app_web/controllers/init_controller.       41        6        0
100.0% lib/app_web/controllers/tag_controller.e       77       25        0
 94.6% lib/app_web/live/app_live.ex                  441      130        7
100.0% lib/app_web/live/stats_live.ex                 77       21        0
100.0% lib/app_web/router.ex                          49        9        0
100.0% lib/app_web/views/error_view.ex                59       12        0
  0.0% lib/app_web/views/profile_view.ex               3        0        0
  0.0% lib/app_web/views/tag_view.ex                   3        0        0
[TOTAL]  98.6%
----------------
Generating report...
```

i.e. test coverage decreased by 1.4% due to the new code. 

Reviewing the `HTML` coverage report we see that the lines that are not covered by tests are exactly the lines that are being added by this PR:

mvp-code-not-covered

Sadly, CI doesn't pass because @panoramix360 created a fork. 
This is a problem with how `GitHub` handles Environment Variables. 🤦‍♂️ 

The only way we can make the CI pass is to make the `AUTH_API_KEY` `public`. 🔓 
Or for @panoramix360 to re-create these changes as a `branch` on the @dwyl source version.
I would prefer to resolve the Environment Variables once and for all to avoid wasting contributor time in the future.

**Comment by nelsonic** [comment]
The CI error is in the Build and Test phase: https://github.com/dwyl/mvp/actions/runs/5514171282/jobs/10060694853?pr=389#step:8:66

image

This is because `GitHub` does not give access to `AUTH_API_KEY` to PRs based on Forks. 
This is a _super_ annoying feature of `GitHub` CI we cannot circumvent. 😢

**Comment by nelsonic** [comment]
To write a test for this new functionality we need to emit a `keydown` event:
https://hexdocs.pm/phoenix_live_view/Phoenix.LiveViewTest.html#render_keydown/1
I'm happy to add this test if you're short on time @panoramix360

**Comment by nelsonic** [comment]
As for the failing Review App Job: https://github.com/dwyl/mvp/actions/runs/5488807164/jobs/10002305167#step:5:20

image

This is exactly the same issue described above, `GitHub` is blocking access to the `FLY_API_TOKEN` environment variable because the PR was created from a `fork`. 

Once this is merged into `main` the deploy to `fly.io` will work fine as it has in the past.

**Comment by nelsonic** [comment]
Sadly, we still have the regression that we cannot add tags dynamically as outlined in: https://github.com/dwyl/mvp/issues/221#issuecomment-1332004568

image

But if I _manually_ add a couple of tags on my `localhost`: 
image

And then test the feature that @panoramix360 is proposing:
image

I cannot get the desired behaviour described by @iteles in https://github.com/dwyl/mvp/issues/308#issue-1587730252 💭

**Comment by nelsonic** [comment]
@iteles could you please help us with the requirements so that we can review this PR ASAP. 🙏

**Comment by nelsonic** [comment]
@panoramix360 meanwhile if you can add a test for the Enter keypress event I will `git pull` from your branch. 🙏

**Comment by panoramix360** [comment]
hey! Thank you for inviting me to the repo, if you prefer I can create the branch on the repo and delete the fork, but as you mentioned, it's good to have that fixed for future contributions.

I can create the tests, no problem, sorry for not creating it earlier, I didn't see that the project had tests.

Regarding the behavior, what happened when you pressed enter? It didn't work?

**Comment by panoramix360** [comment]
hey @nelsonic! do you have any tips on how to test it? I'm new to Phoenix LiveView and I'm thinking of using this one to test it:
https://hexdocs.pm/phoenix_live_view/Phoenix.LiveViewTest.html#render_keydown/1

If you have any ideas or directions to it, it'd be helpful!

**Comment by panoramix360** [comment]
I'm trying to do something like that:
- Create a new tag on the database (`Tag.create_tag`)
- Then I use the `render_keydown` to fire the Enter event
- Then I use the `render_submit` to create a new item on the database
- Then check if the item is created with the first tag

This is my code:
```
test "select tag when enter pressed", %{conn: conn} do
    {:ok, view, html} = live(conn, "/")

    {:ok, _tag1} =
      Tag.create_tag(%{person_id: 0, text: "tag1", color: "#FCA5A5"})

    {:ok, _tag2} =
      Tag.create_tag(%{person_id: 0, text: "enter_tag", color: "#FCA5A5"})

    assert render_keydown(view, "add-first-tag", %{
             "key" => "Enter",
             "value" => "enter_tag"
           })

    assert render_submit(view, :create, %{text: "tag enter pressed"})

    IO.inspect(Tag.list_person_tags(0))
    IO.inspect(Item.get_item!(2).tags)
  end
```

It seems that the `render_keydown` is not finding any tags created, which is strange since I was creating before it.

I believe I'm missing something

**Comment by LuchoTurtle** [comment]
Hey @panoramix360 

Awesome job on proactively working on this, it's mighty appreciated! I've pulled the remote from your forked repo and ran your branch locally and it seems to work as you've described.


https://github.com/dwyl/mvp/assets/17494745/2d34574e-6941-40aa-b030-b5882fab2e17

I've seen you've mentioned it only toggles the tag on top of the list. Although it would make more sense to get the first tag **not selected**, I'm aware that you've tried to make it simple.

When I run the tests `mix c`, they seem to pass as well:

```
Finished in 1.1 seconds (1.0s async, 0.08s sync)
121 tests, 0 failures

Randomized with seed 879703
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/api/item.ex                               218       56        0
100.0% lib/api/tag.ex                                101       24        0
100.0% lib/api/timer.ex                              152       40        0
100.0% lib/app/color.ex                               90        1        0
100.0% lib/app/item.ex                               377       53        0
100.0% lib/app/item_tag.ex                            12        1        0
100.0% lib/app/tag.ex                                108       18        0
100.0% lib/app/timer.ex                              481       90        0
100.0% lib/app_web/controllers/auth_controller.       26        4        0
100.0% lib/app_web/controllers/init_controller.       41        6        0
100.0% lib/app_web/controllers/tag_controller.e       77       25        0
100.0% lib/app_web/live/app_live.ex                  441      130        0
100.0% lib/app_web/live/stats_live.ex                 77       21        0
100.0% lib/app_web/router.ex                          49        9        0
100.0% lib/app_web/views/error_view.ex                59       12        0
  0.0% lib/app_web/views/profile_view.ex               3        0        0
  0.0% lib/app_web/views/tag_view.ex                   3        0        0
[TOTAL] 100.0%
```

Thanks for getting the test coverage back to 100%. I'm glad you've worked more on them (saw your latest commits), the assertions make more sense now and actually test if the tag is selected.

I'm going to make a small comment on the review shortly :)

File: test/app_web/live/app_live_test.exs
Comment (LuchoTurtle) [inline test/app_web/live/app_live_test.exs:830]: Good test :) 

You could take it just a *tiny bit* step further and actually **create** the item. Then, call `Item.get_item(id, true)` (true will preload the tags) and `assert` if the `tag` you've created is associated with it.

After that, it's absolutely perfect 🤩

**Review by LuchoTurtle — CHANGES_REQUESTED** [review]

File: test/app_web/live/app_live_test.exs
Comment (panoramix360) [reply test/app_web/live/app_live_test.exs:830]: OK! I'll do this, but how can I know the id of the item? since it was created by the `render_submit`?

Comment (LuchoTurtle) [reply test/app_web/live/app_live_test.exs:830]: Since `render_submit` only returns the view, the easiest way would be to use `Item.list_items` and find the item with the same text you've created. Because each unit test is contained, there shouldn't be any performance issues :)

Comment (panoramix360) [reply test/app_web/live/app_live_test.exs:839]: Done! I did a little different solution, let me know what you think!

**Review by LuchoTurtle — APPROVED** [review]
Looks good! 
I'll let @nelsonic have a final say.

Thanks for the PR 😄

**Review by nelsonic — APPROVED** [review]
@panoramix360 great contribution. 🎉 
Thanks for adding the tests. ✅ 

Confirmed on `localhost`:

```sh
git pull
mix c

Finished in 1.3 seconds (1.2s async, 0.08s sync)
121 tests, 0 failures

Randomized with seed 811468
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/api/item.ex                               218       56        0
100.0% lib/api/tag.ex                                101       24        0
100.0% lib/api/timer.ex                              152       40        0
100.0% lib/app/color.ex                               90        1        0
100.0% lib/app/item.ex                               377       53        0
100.0% lib/app/item_tag.ex                            12        1        0
100.0% lib/app/tag.ex                                108       18        0
100.0% lib/app/timer.ex                              481       90        0
100.0% lib/app_web/controllers/auth_controller.       26        4        0
100.0% lib/app_web/controllers/init_controller.       41        6        0
100.0% lib/app_web/controllers/tag_controller.e       77       25        0
100.0% lib/app_web/live/app_live.ex                  441      130        0
100.0% lib/app_web/live/stats_live.ex                 77       21        0
100.0% lib/app_web/router.ex                          49        9        0
100.0% lib/app_web/views/error_view.ex                59       12        0
  0.0% lib/app_web/views/profile_view.ex               3        0        0
  0.0% lib/app_web/views/tag_view.ex                   3        0        0
[TOTAL] 100.0%
----------------
Generating report...
Saved to: cover/
```

```
mix s
```

dwyl-mvp-enter-tags

@LuchoTurtle thanks for reviewing + trouble-shooting. ❤️

**Pull request merged**
